compiler/ast/checked_ast.nim

```diff
@@ -0,0 +1,112 @@
+## The ``CheckedAst`` type is implemented here, along with its accessor
+## routines and routines for creating and manipulating ``CheckedAst``
+## instances.
+##
+## ``CheckedAst`` is a wrapper around ``PNode``, and is intended to make
+## propagation less manual, by keeping track of whether there is an error
+## *somewhere* in the managed AST.
+##
+## The idea is to use ``CheckedAst`` as the return type of all semantic
+## analysis procedures, eventually removing the need for wrapper errors
+## (``adWrappedError``).
+
+import
+  compiler/ast/[
+    ast,
+    ast_types,
+    errorhandling,
+    lineinfos
+  ],
+  compiler/front/[
+    options
+  ]
+
+type
+  CheckedAst* = object
```

Comment:
```suggestion
  TrackedAst* = object
```

We're more tracking than checking.  I think checking would be something like assertions on replacing an error node with something else.

Reply:
Adding the bit about `sem-checked` as the etymology of the name in a doc comment is also workable, sorta recontextualized it for me. Your call, this is one of those things where can easily change our minds later.

Reply:
Okay, given that the module already uses the name, I'd leave it as-is for the time being (but with an added explained for the name in the doc comment).